onchain/src/bip340.cairo

```diff
@@ -0,0 +1,364 @@
+use core::result::ResultTrait;
+use core::to_byte_array::AppendFormattedToByteArray;
+use core::byte_array::ByteArrayTrait;
+use core::traits::Into;
+use core::option::OptionTrait;
+use core::starknet::SyscallResultTrait;
+use core::sha256::compute_sha256_byte_array;
+use core::debug::print_byte_array_as_string;
+
+// bip340 implementation
+// references:
+//   Schnorr signatures explained:
+//   https://www.youtube.com/watch?v=wjACBRJDfxc&ab_channel=Bitcoinology
+//   NIP-01:
+//   https://github.com/nostr-protocol/nips/blob/master/01.md
+//   BIP-340:
+//   https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki
+//   reference implementation:
+//   https://github.com/bitcoin/bips/blob/master/bip-0340/reference.py
```

Comment:
put that on top as module documentation
https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html